### PR TITLE
rewrite jax.checkpoint (aka jax.remat), leave old implementation for now

### PR DESCRIPTION
**Pull request by mattjj**
The new `jax.checkpoint` is currently engaged whenever we pass an explicit policy. To get a "remat everything" decorator, one should write `@partial(jax.checkpoint, policy=lambda *_, **__: False)`.

Because the new code path is used whenever an explicit policy is passed, that means that all our parameterized remat tests already cover this new code path. (And I made sure it's covered, by putting an `assert False` in the new implementation and running the old tests!)

Fundamentally `jax.checkpoint` is about customizing the partial evaluation involved in autodiff. So our previous implementation was just a call primitive with a custom partial evaluation rule, inheriting most of its behaviors from the other call-processing machinery. (We also need a custom XLA translation rule, to foil CSE, and a custom transpose rule, to handle the nonlinear terms during transposition.)

But because JAX's partial evaluation works by data dependence only, that meant constants were "invisible" to `jax.checkpoint`, in the sense that they would unavoidably be evaluated / constant-folded and not recomputed.

It turns out that this behavior could be both surprising and costly. Here's a toy example where a potentially-large constant can be created behind the scenes:

```python
import jax
import jax.numpy as jnp

# utility for inspecting the types of residuals
def saved_residuals(f, *args, **kwargs):
  linearize = lambda *args: jax.linearize(f, *args, **kwargs)[1]
  jaxpr = jax.make_jaxpr(linearize)(*args).jaxpr
  return [v.aval for v in jaxpr.outvars]

# example function. no residuals should be stored, right? ...right?
@jax.checkpoint
def f(x):
  return jnp.einsum('ii->i', x)  # extract the diagonal

print(saved_residuals(f, jnp.ones((2, 2))))  # [ShapedArray(float32[2,2]), ShapedArray(float32[2,2])]
```

The reason residuals are being saved here is that to extract diagonals `jnp.einsum` basically creates an identity matrix, more or less evaluating `jnp.einsum('ij,ij->i', jnp.eye(N), x)`. And `jax.checkpoint` would save that identity matrix as a residual!

This toy example doesn't look bad, but actually in large transformer layers this could lead to saving constants which scaled like the square of the sequence length: `logits: f32[B, N, T, T] = einsum('BTNH,BTNH->BNTT', q_proj, k_proj) / sqrt(H)`. Moreover, these constants need not be materialized at all: XLA can often fuse their creation into consumers!

To fix this, we want `jax.checkpoint` to be able to see constant creation, and not just see operations from data dependence. Sounds like omnistaging, #3370! Given JAX's current tracing machinery, I couldn't think of a way to tweak the existing partial evaluation rule to get what we wanted. But it was straightforward just to adapt `jax.checkpoint` into an "initial-style primitive" like the control flow primitives. (See Autodidax for some explanation of that jargon.) That "initial-style" implementation meant we have to write a bit more explicit boilerplate-y rules, but they weren't so bad. And we gain a big upside!

Another limitation of the new implementation is that it can't handle `concrete=True`. I think that we can just get rid of that option; it seems no one is using it inside Google. (If some OSS people need it, we can figure out some kind of workaround when the need arises...)

This enables us to fix several other issues, so I've wanted to do it for a while:
1. it gives us the tools to properly handle effectful primitives in `jax.checkpoint` (which might mean just raising an error, rather than silently failing like today), since otherwise those lack data dependence and hence have the same 'constant folding' issue;
2. it lets `jax.checkpoint` avoid a [`scan` optimization](https://github.com/google/jax/blob/d703def10406097abd80853a0b64a991e752739d/jax/_src/lax/control_flow.py#L1625) which can hoist creation of constants outside a loop, defeating consumer fusions (whereas previously using `jax.checkpoint` would not help there).

I left the old implementation in place mainly so that we can land this in stages.

### CODE REVIEW

**Review by zhangqiaorjc — APPROVED** [review]
Thanks Matt! Will be immensely helpful for our use case.

**Review by jekbradbury — APPROVED** [review]

**Comment by shoyer** [comment]
Sounds good, I'll ping you about this when it comes up! We will likely be getting started on this in a month or two.
